numpy/core/shape_base.py

```diff
@@ -458,47 +458,37 @@ def _concatenate_shapes(shapes, axis):
     return first_shape[:axis] + (shape_on_dim,) + first_shape[axis+1:]
 
 
-def _shape_recursion(arrays, list_ndim, result_ndim, depth=0):
-    """Internal implementation that gets the shape of the final array.
-    """
-    if depth < list_ndim:
-        shapes = [_shape_recursion(arr, list_ndim, result_ndim, depth + 1)
-                  for arr in arrays]
-        concat_depth = depth + result_ndim - list_ndim
-        return _concatenate_shapes(shapes, concat_depth)
-    else:
-        # We've 'bottomed out'
-        return arrays.shape
-
-
 def _atleast_nd(a, ndim):
     # Ensures `a` has at least `ndim` dimensions by prepending
     # ones to `a.shape` as necessary
     return array(a, ndmin=ndim, copy=False, subok=True)
 
 
-def _dtype_recursion(arrays, list_ndim, depth=0):
-    """Internal implementation that gets the dtype of the final array.
-    """
-    if depth < list_ndim:
-        dtypes = [_dtype_recursion(arr, list_ndim, depth + 1)
-                  for arr in arrays]
-        return _nx.result_type(*dtypes)
-    else:
-        # We've 'bottomed out'
-        return arrays.dtype
+def _info_recursion(arrays, list_ndim, result_ndim, depth=0):
+    """From the input arrays to block, return a tuple containing
 
-
-def _to_uniform_dim_recursion(arrays, list_ndim, result_ndim, depth=0):
-    """Convert all inputs to arrays."""
+    * a list of normalized arrays using _atleast_nd,
+    * the resulting shape of the final array
+    * the resulting dtype of the final array
+    """
     if depth < list_ndim:
         if len(arrays) == 0:
             raise ValueError('Lists cannot be empty')
-        return [_to_uniform_dim_recursion(arr, list_ndim, result_ndim,
-                                          depth + 1)
+        concat_depth = depth + result_ndim - list_ndim
+        info = [_info_recursion(arr, list_ndim, result_ndim, depth + 1)
                 for arr in arrays]
+        arrays, shapes, dtypes = zip(*info)
+        # offsets = (0,) + tuple(
+        #     _nx.cumsum([shape[concat_depth] for shape in shapes]))[:-1]
```

Comment:
cumsum will be slower than writing out the loop I had elsewhere in a comment, I think